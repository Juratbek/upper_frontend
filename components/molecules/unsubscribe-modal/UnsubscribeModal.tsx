import { Clickable, Modal, Spinner } from 'components/lib';
import { FC, useCallback, useMemo } from 'react';
import { useDispatch, useSelector } from 'react-redux';
import { useUnsubscribe } from 'store/clients/subscription';
import {
  closeUnsubscribeModal,
  getBlogIdToUnsubscribe,
  getIsUnsubscribeModalOpen,
} from 'store/states';

import classes from './UbsubscribeModal.module.scss';

export const UnsubscribeModal: FC = () => {
  const isOpen = useSelector(getIsUnsubscribeModalOpen);
  const dispatch = useDispatch();
  const blogId = useSelector(getBlogIdToUnsubscribe);
  const { mutate: unsubscribe, isPending: isUnsubscribeLoading } = useUnsubscribe(Number(blogId));

  const closeHandler = useCallback(() => dispatch(closeUnsubscribeModal()), []);

  const unsibscribeHandler = useCallback(() => {
    unsubscribe('');
    closeHandler();
  }, [closeHandler]);

  const footer = useMemo(
    () => (
      <div className='d-flex'>
        <Clickable className={classes['unsubscribe-btn']} onClick={unsibscribeHandler}>
          {isUnsubscribeLoading ? <Spinner /> : 'Bekor qilish'}
        </Clickable>
        <Clickable className={classes['no-btn']} onClick={closeHandler}>
          yo&apos;q
        </Clickable>
      </div>
    ),
    [closeHandler, isUnsubscribeLoading, unsibscribeHandler],
  );

  return (
    <Modal isOpen={isOpen} close={closeHandler} footer={footer}>
      <p className={classes.text}>Obunani bekor qilmoqchimisiz?</p>
    </Modal>
  );
};
