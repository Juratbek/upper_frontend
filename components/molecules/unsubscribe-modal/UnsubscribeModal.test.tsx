import { render, screen, waitFor } from '@testing-library/react';
import userEvent from '@testing-library/user-event';
import { QueryWrapper, ReduxWrapper } from 'tests';
import { vi } from 'vitest';

import { UnsubscribeModal } from './UnsubscribeModal';

const mocks = vi.hoisted(() => ({
  closeUnsubscribeModal: vi.fn().mockReturnValue({ type: 'CLOSE_MODAL' }),
  useUnsubscribe: vi.fn().mockReturnValue({ mutate: vi.fn() }),
}));

vi.mock('store/states', async (importOriginal) => {
  const original = await importOriginal<typeof import('store/states')>();

  return {
    ...original,
    closeUnsubscribeModal: mocks.closeUnsubscribeModal,
  };
});

vi.mock('store/clients/subscription', () => ({
  useUnsubscribe: mocks.useUnsubscribe,
}));

const Component = (
  <ReduxWrapper>
    <QueryWrapper>
      <UnsubscribeModal />
    </QueryWrapper>
  </ReduxWrapper>
);

describe('Unsubscribe modal', () => {
  it('snapshot', () => {
    render(Component);
    expect(document.body).toMatchSnapshot();
  });

  it('closes the modal on cancel click', async () => {
    render(Component);

    const cancelButton = screen.getByRole('button', { name: "yo'q" });
    userEvent.click(cancelButton);

    await waitFor(() => {
      expect(mocks.closeUnsubscribeModal).toHaveBeenCalledTimes(1);
    });
  });

  it('unsubscribes and closes the modal on unsubscribe', async () => {
    const mockUnsubscribe = vi.fn();
    mocks.useUnsubscribe.mockReturnValue({ mutate: mockUnsubscribe });
    render(Component);

    const unsubscribeButton = screen.getByRole('button', { name: 'Bekor qilish' });
    userEvent.click(unsubscribeButton);

    await waitFor(() => {
      expect(mocks.closeUnsubscribeModal).toHaveBeenCalled();
      expect(mockUnsubscribe).toHaveBeenCalledTimes(1);
    });
  });

  it('shows spinner if unsubscribe is loading', () => {
    mocks.useUnsubscribe.mockReturnValue({ mutate: vi.fn(), isPending: true });
    render(Component);

    expect(document.body).toMatchSnapshot();
  });
});
