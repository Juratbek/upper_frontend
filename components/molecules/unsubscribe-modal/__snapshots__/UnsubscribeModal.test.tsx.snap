// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Unsubscribe modal > shows spinner if unsubscribe is loading 1`] = `
<body>
  <div>
    <div
      class="modal"
      data-testid="modal-root"
    >
      <button
        class="root modal-bg"
      />
      <div
        class="modal-dialog modal-dialog--undefined"
      >
        <div
          class="modal-body"
        >
          <p
            class="text"
          >
            Obunani bekor qilmoqchimisiz?
          </p>
        </div>
        <div
          class="d-flex"
        >
          <button
            class="root unsubscribe-btn"
          >
            <div
              class="spinner spinner--primary"
              role="spinner"
            />
          </button>
          <button
            class="root no-btn"
          >
            yo'q
          </button>
        </div>
      </div>
    </div>
  </div>
</body>
`;

exports[`Unsubscribe modal > snapshot 1`] = `
<body>
  <div>
    <div
      class="modal"
      data-testid="modal-root"
    >
      <button
        class="root modal-bg"
      />
      <div
        class="modal-dialog modal-dialog--undefined"
      >
        <div
          class="modal-body"
        >
          <p
            class="text"
          >
            Obunani bekor qilmoqchimisiz?
          </p>
        </div>
        <div
          class="d-flex"
        >
          <button
            class="root unsubscribe-btn"
          >
            Bekor qilish
          </button>
          <button
            class="root no-btn"
          >
            yo'q
          </button>
        </div>
      </div>
    </div>
  </div>
</body>
`;
